playbooks/kubernetes/app-resource-tuning.md

```diff
@@ -0,0 +1,111 @@
+Kubernetes has knobs that allow you to allocate cpu/memory for an app.
+
+These settings are configured in k8s manifests (Hokusai staging/production.yml), in each app's repo. Hokusai templates come with example values for these settings:
+
+https://github.com/artsy/artsy-hokusai-templates
+
+However, we should tune them per app. This doc provides some guidelines on tuning.
+
+
+# Container CPU/Memory Request/Limit.
+
+https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/
+
+https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/
+
+These parameters are set per container. `requests` means how much of a resource will be reserved for the container. `limits` means the maximum amount of a resource that the container is allowed to use. If an app hits CPU limit, it will get throttled and slows down but it won't terminate. If it hits memory limit, it will likely terminate with out-of-memory error.
+
+CPU values are expressed in "number of CPU's". For example, 0.2 means 0.2 cpus. The value can also be expressed in units of milli cpus. So 200m is also 0.2 cpus. We prefer to use milli cpu units.
+
+For memory, the value is in bytes.
+
+For a pod that has a web container as well as an nginx container, our practice is that we set these parameters for the web container only.
+
+```
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: example-web
+  namespace: default
+spec:
+...
+  template:
+...
+    spec:
+      containers:
+        - name: example-web
+
+          resources:
+            requests:
+              cpu: ???
+              memory: ???
+            limits:
+              cpu: ???
+              memory: ???
+```
+
+## General Recommendations
+These are very general recommendations, as CPU and memory usage are both hard to gauge. Some apps such as Pulse use a lot of CPU but only momentarily. For memory, some apps such as Gemini try to use more and more until they hit OOM and crash.

```

Comment:
s/until they hit OOM and crash/until they hit OOM and are killed/

Reply:
agreed.